### ISSUE TITLE
Alternative Storage Backends

### ISSUE THREAD

**Issue by reallyquickturtle** [pypiserver/pypiserver#253]
Allow to store packages in cloud storage services such as AWS S3. 

This can be particularly useful when deploying on cloud infrastructure that has no persistent filesystem available (ECS, Heroku, etc.)

**Comment by tomsaunders-code** [comment]
@mplanchard or @ankostis is this something you would be interested in supporting? I think may be able to help contribute to here.

The NPM community has a really nice private package server called Verdaccio that features this kind of plugin based storage backend.

https://github.com/verdaccio/verdaccio
https://github.com/Remitly/verdaccio-s3-storage

**Comment by jayvdb** [comment]
I would be interested in a local backend that re-uses the real pip cache. c.f. https://github.com/jerith/landinggear/pull/2

**Comment by deemetree-at-satelligence** [comment]
If anyone would be willing to collaborate and kick it off I would be happy to join up as well, looking in particular for options to conveniently use it with Google Cloud Storage in App Engine. (Possibly I have misread or not found a solution that already exists)

**Comment by mplanchard** [comment]
This is definitely something we'd like to support! It's one of the major things I'd like to get into pypiserver this year.

**Comment by deemetree-at-satelligence** [comment]
@mplanchard I have created a simple solution for the time being as a plugin to manage files between App Engine and Cloud Storage and would then submit it as a pull request if you are interested! Would much appreciate comments and collaboration on that further.

**Comment by mplanchard** [comment]
@deemetree-at-satelligence yep, I'd love to work on incorporating that into pypiserver and am happy to collaborate w/you on that! If more discussion is needed before opening a PR, feel free to open a new issue to discuss it, or go ahead and open a PR if you'd like 🎉

**Comment by deemetree-at-satelligence** [comment]
@mplanchard sorry, was a bit swamped this month with other tasks, but am picking it up! I will open up a WIP pull request soon

**Comment by mplanchard** [comment]
Hey, no rush at all. I'm in the same boat 99.9% of the time 😄

**Comment by Luttik** [comment]
Any news on this? I would love this functionality.

**Comment by mplanchard** [comment]
- @Luttik not sure where the others are at, but this is far from a simple thing to implement. Particularly with incoming changes that require unzipping and introspecting package metadata to provide the "requires-python" data for pip (see #320), we're going to need to figure out some kind of robust caching strategy to support remote stores with anything like acceptable latency. This of course becomes much more complex when you consider that a remote store allows multiple instances of pypiserver to hit it simultaneously. 

I've got some long term ideas I'm slowly working on in between other side projects that should help move us in this direction, and it's possible we'll get an acceptable PR to add in something like this before I can get around to it, but if you need this in the short term, you should probably check out other package managers.

I do still think it's not unreasonable that we might get it done this year sometime, but definitely not soon.

**Comment by Luttik** [comment]
@mplanchard thanks for the info that is really insightful. Maybe we can learn something from PyPiCloud in this regard. Since they have already implemented bucket support. (PyPiCloud has other issues though; mediocre documentation, an unintuitive interface and unexpected errors or inconsistencies).

**Comment by deemetree-at-satelligence** [comment]
@mplanchard @Luttik hello! sorry for my afk for a while, but coincidentally I am picking it up from the planned list and started working on finalizing it again as of last week. I am planning to be ready with a some minimal example integrated to Pypiserver next week, I forked the project here (it's been a bit stale for a while...) [https://github.com/deemetree-at-satelligence/pypiserver/](https://github.com/deemetree-at-satelligence/pypiserver/tree/deemetree-appengine-plugin), deemetree-appengine-plugin branch. I am not yet sure how to proceed with opening the PR and all, but will take a look at CONTRIBUTING or other pull requests for inspiration :) 

As a summary of the current idea is: synchronize files on pre and post processing of the requests.  
1. `AppEngine` has only [`/tmp` storage](https://cloud.google.com/appengine/docs/standard/python3/using-temp-files) that is cleaned on each instance. 
2. `Google Storage` allows to host any files in buckets, so just as well could be used to store packages.
3. Since `pypiserver` is based on `bottle`, we can use request [`hooks`](https://bottlepy.org/docs/dev/recipes.html#using-the-hooks-plugin) to synchronize the files locally and remotely.  

There can be issues with this setup, buuuut it is working decently for us for a while now (under a rather small amount of requests). Actually would appreciate discussion on this idea.

**Comment by deemetree-at-satelligence** [comment]
I opened a separate issue here: https://github.com/pypiserver/pypiserver/issues/322, let me know if anything should be adjusted
